### PR TITLE
engine_pytest

### PR DESCRIPTION
**Pull request by MariamTurky**

### CODE REVIEW

File: test_engine.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    Parameters: engine(object) : An instance of the Engine class.
+
+    Returns: None
+    """
+    opt = engine.get_optimizer('SGD', lr=0.01, momentum=0.9)
+    assert isinstance(opt, torch.optim.SGD)
+    assert opt.defaults['lr'] == 0.01
+    assert opt.defaults['momentum'] == 0.9
+    
+def test_get_scheduler_steplr(engine):
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:97]: This way, you are testing each getter function independently. Maybe you should test their functionality within the engine itself. For example, make sure that upon setting the configs and creating an engine instance, its attributes will carry the expected values returned from the corresponding getters.

File: test_engine.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    Description: Tests the get_metrics method of the engine with the "dice" metric.
+
+    Parameters: engine(object) : An instance of the Engine class.
+
+    Returns: None
+    """
+    metric = engine.get_metrics('dice')
+    assert isinstance(metric, DiceMetric)
+
+def test_get_metrics_jaccard(engine):
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:143]: Try to combine these test cases in one test. Applicable also to the previous tests.

File: test_engine.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                    ForegroundMaskD(
+                        Keys.LABEL,
+                        threshold = 0.5, 
+                        invert = True, 
+                        allow_missing_keys = True
+                    ),
+                    ToTensorD(Keys.all(), allow_missing_keys = True),
+                ]
+            )
+    engine.postprocessing_transforms = Compose([
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:221]: stick to the indentation and spacing convention

File: test_engine.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+    assert len(engine.train_dataloader) > 0
+    assert len(engine.test_dataloader) > 0
+    assert isinstance(engine.train_dataloader, MonaiLoader)
+    assert isinstance(engine.val_dataloader, MonaiLoader)
+    assert isinstance(engine.test_dataloader, MonaiLoader)
+
+
+#########################################################################################################################
+def test_save_checkpoint(engine,ckpt_path='Checkpoint'):
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:247]: Maybe you can test the loading function first, then make sure that saving and reloading a checkpoint yield the same weights.

File: test_engine.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    init_weights = engine.network.state_dict()
+
+    engine.load_checkpoint(checkpoint_path)
+
+    # Get loaded weights
+    loaded_weights = engine.network.state_dict()
+
+    # Check that weights match
+    for i in init_weights.keys():
+        assert torch.allclose(init_weights[i], loaded_weights[i])
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:288]: Won't the initial weights be random? how do you expect them to be close the checkpoint?

File: test_engine.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+#########################################################################################################################
+
+def test_fit(engine,
+        epochs = 2,
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:292]: How do you pass these arguments to the test?

File: test_engine.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    engine.fit(
+        epochs = epochs,
+        evaluate_epochs = evaluate_epochs,
+        batch_callback_epochs = batch_callback_epochs,
+        save_weight = save_weight,
+        save_path = save_path
+    )
+
+    assert os.path.exists(save_path)
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:329]: testing a training function should be more complex that simply asserting that it saves the results. Either assert that the trained weights themselves are close to a reference cp (after setting a seed), or maybe you can assert that the progress (loss and metric) is close to an expected reference.

File: test_engine.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+    Returns: None
+    """
+
+    get_transforms(engine)
+    engine.load_data()
+
+    loss, metric = engine.test()
+
+    assert loss != 0
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:348]: set a seed, then assert for the actual values

File: test_engine.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+    Returns: None
+    """
+    
+    get_transforms(engine)
+    engine.load_data()
+
+    predicted = engine.predict(path)
+
+    assert predicted.shape == torch.Size([1,1,256,256,128])
```
Comment (Muhammad-Al-Barbary) [inline test_engine.py:370]: Check that the values of the prediction are similar to a reference.

**Review by Muhammad-Al-Barbary — COMMENTED** [review]
Good job! 
A general comment is to always stick to the convention. Also, putting in proper effort to develop these tests will greatly facilitate future development, so please strive to do your best.

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (MariamTurky) [reply .vscode/settings.json:1]: done

File: test_engine.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    ToTensorD,
+    RandAdjustContrastd,
+    RandZoomd,
+    ActivationsD,
+    AsDiscreteD,
+)
+#######################################################################################################################
+def set_configs():
+    """
+        Description: Sets new values for config parameters.
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 7 above)
Comment (MariamTurky) [reply test_engine.py:296]: done

File: test_engine.py (hunk 8 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 9 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 6 above)
Comment (MariamTurky) [reply test_engine.py]: i have fixed this

File: test_engine.py (hunk 5 above)
Comment (MariamTurky) [reply test_engine.py]: i have fixed this

File: test_engine.py (hunk 4 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 3 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 2 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 1 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+    engine.load_data()
+
+    # Get initial weights
+    init_weights = [p.clone() for p in engine.network.parameters()]
+
+    # Get loaded weights
+    engine.load_checkpoint(ref_path)
+    ref_weights = engine.network.parameters()
+
+    engine.fit(save_path=save_path)
```
Comment (helrewaidy) [inline test_engine.py:294]: how many epochs this fit will run? please set it to 1
Also, the fit function by default doesn't save the weights (not sure why :) ) so please enable save_weights param

File: test_engine.py [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    ----------
+        None
+    """
+
+    get_transforms(engine)
+    engine.load_data()
+
+    predicted = engine.predict(path)
+    prediction = predicted.cpu()
+    prediction = prediction.numpy()
```
Comment (helrewaidy) [inline test_engine.py:353]: you can make it in one step: `prediction = predicted.cpu().numpy()`

**Review by helrewaidy — COMMENTED** [review]
Great job, Mariam. I've left few comments for you. Also please add the test cases that you've used. one or two samples are enough. Thank you

File: tests/testdata/train/volume/resized_liver.nii
Comment (Muhammad-Al-Barbary) [inline tests/testdata/train/volume/resized_liver.nii:1]: instead of duplicating the data for training and testing, you can save it once only and set the training and testing paths in set_configs to the same path.

File: tests/testdata/reference_engine_cp
Comment (Muhammad-Al-Barbary) [inline tests/testdata/reference_engine_cp:1]: Maybe we can categorize the files in subfolders, e.g. checkpoints, testcases.... Only a suggestion tho.

File: tests/testdata/engine_cp
Comment (Muhammad-Al-Barbary) [inline tests/testdata/engine_cp:1]: I think we should add a '.pt' extension to all checkpoints, for clarification.

**Review by Muhammad-Al-Barbary — COMMENTED** [review]
Great work! The tests are improving rapidly. 
All my comments are suggestions and open to discussion.

File: .gitignore [hunk 13]
```diff
@@ -29,3 +29,4 @@ lobe_cp
 mri_cp
 spleen_cp
 lobe_cp_sliding_window
+*.json
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply .gitignore]: done

File: test_engine.py [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        engine: object
+            An instance of the Engine class.
+
+    Returns
+    ----------
+        None
+    """
+
+    assert isinstance(engine.optimizer, torch.optim.Adam)
+    assert engine.optimizer.defaults["lr"] == 0.01
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+        engine: object
+            An instance of the Engine class.
+
+    Returns
+    ----------
+        None
+    """
+
+    get_transforms(engine)
+    engine.load_data()
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: first, I give the path of the data to set_configs(),
ex: config.dataset["training"] = "tests/testdata/data" 
then, the data is being loaded inside load_data().
ex: trainloader = DataLoader(
            dataset_path = config.dataset["training"],)

File: test_engine.py [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+
+    assert len(engine.train_dataloader) > 0
+    assert len(engine.test_dataloader) > 0
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+    assert isinstance(engine.train_dataloader, MonaiLoader)
+    assert isinstance(engine.val_dataloader, MonaiLoader)
+    assert isinstance(engine.test_dataloader, MonaiLoader)
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+    """
+
+    assert os.path.exists(ckpt_path)
+
+    loaded_checkpoint = torch.load(ckpt_path)
+
+    engine.load_checkpoint(ckpt_path)
+
+    # verify network state dict match
+    assert set(engine.network.state_dict()) == set(loaded_checkpoint["state_dict"])
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+    ref_weights = engine.network.state_dict()
+
+    engine.load_checkpoint(checkpoint_path)
+
+    # Get loaded weights
+    loaded_weights = engine.network.state_dict()
+
+    # Check that weights match
+    for i in ref_weights.keys():
+        assert torch.allclose(ref_weights[i], loaded_weights[i], atol=1e-3)
```
[the remark this answers is not in the document]
Comment (MariamTurky) [reply test_engine.py]: done

File: test_engine.py (hunk 11 above)
Comment (MariamTurky) [reply test_engine.py]: done

File: tests/test_engine.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+
+    # Load checkpoint
+    loaded_checkpoint = torch.load(ckpt_path)
+
+    # Get loaded weights
+    engine.load_checkpoint(ckpt_path)
+    loaded_weights = engine.network.parameters()
+
+    # verify network state dict match
+    assert set(engine.network.state_dict()) == set(loaded_checkpoint["state_dict"])
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:272]: why are you casting the weights to a set?

File: tests/test_engine.py [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+    ----------
+        None
+    """
+    # Path of the input directory
+    ckpt_path = config.save["engine_checkpoint"]
+
+    # Get initial weights
+    init_weights =[p.clone() for p in engine.network.parameters()]
+
+    assert os.path.exists(ckpt_path)
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:259]: do you need a reference cp? aren't you overwriting it in the next line?

File: tests/test_engine.py [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+        assert np.allclose(train_batch["label"].numpy(), expected_label, 1e-2)
+
+    assert len(engine.train_dataloader) == 1
+    assert len(engine.test_dataloader) == 1
+    assert isinstance(engine.train_dataloader, MonaiLoader)
+    assert isinstance(engine.val_dataloader, MonaiLoader)
+    assert isinstance(engine.test_dataloader, MonaiLoader)
+
+
+def test_save_load_checkpoint(engine):
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:240]: This function's logic is not straight-forward, maybe you can divide the function into two parts (with comments), one for testing the save function and another for testing the load one, in a way that they are separate but complements each other.

File: tests/test_engine.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+    # Get initial weights
+    init_weights = [p.clone() for p in engine.network.parameters()]
+
+    engine.fit(epochs=1, save_weight=True, save_path=save_path)
+
+    assert os.path.exists(save_path)
+
+    # Verify weights were updated
+    for p0, p1 in zip(init_weights, engine.network.parameters()):
+        assert not torch.equal(p0, p1)
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:309]: be more precise and compare with a reference cp.

**Review by Muhammad-Al-Barbary — COMMENTED** [review]
When addressing a comment, please make sure you apply the necessary modifications to other parts as well.

File: tests/test_engine.py [hunk 24]
```diff
@@ -252,31 +253,36 @@ def test_save_load_checkpoint(engine):
...
+    # Path of reference checkpoint
+    ref_path =  config.save["reference_checkpoint"]
 
     assert os.path.exists(ckpt_path)
+    assert os.path.exists(ref_path)
 
-    # Save checkpoint
-    engine.save_checkpoint(ckpt_path)
-
+    # Test for saving checkpoint
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:262]: you are not even using the save function here, how are you testing it?
The logic of this function should be something like this (pseudocode):

\# Test load
torch_loaded_weights = torch.load(reference_load_cp)
engine.load(reference_load_cp)
engine_loaded_weights = engine.network.get_weights()
assert torch_loaded_weights == engine_loaded_weights # if true, then load function is working properly.

\# Test save
initial_weights = engine.network.get_weights()
assert not exists(saved_cp)
engine.save(saved_cp)
assert exists(saved_cp) #make sure a cp is saved
engine.load(saved_cp) # at this point, we know that load is functioning properly
saved_weights = engine.network.get_weights()
assert saved_weights == initial_weights #if they are not equal, and load is functioning properly, then save is not
delete(saved_cp)

File: tests/test_engine.py [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+    init_weights =[p.clone() for p in engine.network.parameters()]
+
+
+
+    # Test for saving checkpoint
+   
+    assert not os.path.exists("saved_cp.pt")
+    engine.save_checkpoint("saved_cp.pt")
+    assert os.path.exists("saved_cp.pt")
+    engine.load_checkpoint("saved_cp.pt")
```
Comment (Muhammad-Al-Barbary) [inline tests/test_engine.py:269]: still not straight-forward and I think is prone to bugs.

**Review by Muhammad-Al-Barbary — APPROVED** [review]
Several comments are unaddressed, but I will approve anyways assuming that you tested these tests (when everything is fine they pass, and when functions are buggy or references are changed they fail).
If you didn't, you should test them, and please try to address the comments before merging.
